doc/eip-draft/eip-draft_fair_atomic_swap.md

```diff
@@ -0,0 +1,824 @@
+
+The fair Atomic Swap Smart Contract should follows the syntax and semantics of 
+the stateful smart contract in Ethereum, with hash locks support and time locks support.
+
+## Rationale
+<!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->
+
+To resolve the "Free Option Problem", the unfair behaviour should receive punishment.
+
+To apply the punishment, it requires that, besides the asset used to exchange, the initiator must put some asset as collateral, which is in fact premium. The transaction for the premium needs to be locked with the same secret hash but with a flipped payout, i.e. when redeemed with the secret, the premium goes back to the initiator and after timelock, the premium goes to the participant as a compensation for initiator not revealing the secret.
+
+However, this introduces a new problem: this time the participant gains the optionality -- it can get the premium without paying anything, by never participating in.
+
+To resolve the new problem, the premium:
+
++ should be refunded back to the initior if the participant does not participate in at all, or if **the participant's funding is redeemed by the initior**; or
++ should be redeemed for the participant if the initior holds the secret used in hash lock maliciously, which also implies that **the participant's funding is refunded back**.
+
+Such a logic is hard to implement in stateless script system like Bitcoin, because, in the aspect of transaction, **where the premium should go, strictly depends on where the participant's funding goes**, but can be empowered in Ethereum-style stateful systems.
+
+In a word, a simple HTLCs-based Atomic Swap on Spot scenario is natively equivalent to an American Call Option without premium, but adding well-designed premium mechanism mitigates the arbitrage risk for both parties.
+
+Meanwhile, it is also worthy to investigate on HTLCs-based Atomic Swaps on American Call Option scenario, with built-in premium.
+
+In American Call Option with premium, the premium:
+
++ should be refundable for the initior if the participant doesn't participate; or
++ should be redeemable for the participant if the participant does participate.
+
+## Backwards Compatibility
+<!--All EIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The EIP must explain how the author proposes to deal with these incompatibilities. EIP submissions without a sufficient backwards compatibility treatise may be rejected outright.-->
```

Comment:
Keep this section, but remember to change terms like `contract` -> `token standard`